From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Tue, 6 Oct 2026 07:10:00 +0200
Subject: [PATCH] Add DS buttons

in_nds.c sends the keys of Quake's default bindings for the buttons in
game and the menu keys everywhere else, so they can still be rebound. A
tap on a weapon of the touch screen (sbar_nds.c) selects it.
pmMainLoop runs with every key poll: closing the lid puts the DS to
sleep, the DSi power button or L+R+START+SELECT quit, keeping the config
like the quit command.
---
 WinQuake/in_nds.c | 126 ++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 126 insertions(+)
 create mode 100644 WinQuake/in_nds.c

diff --git a/WinQuake/in_nds.c b/WinQuake/in_nds.c
new file mode 100644
index 0000000..abb69c2
--- /dev/null
+++ b/WinQuake/in_nds.c
@@ -0,0 +1,126 @@
+/*
+Copyright (C) 1996-1997 Id Software, Inc.
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; either version 2
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+
+See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+
+*/
+// in_nds.c -- Nintendo DS buttons and touch screen
+//
+// The buttons send the keys of Quake's default bindings in game and the menu
+// keys everywhere else, so they can still be rebound. A tap on a weapon of
+// the touch screen (sbar_nds.c) selects it.
+
+#include "quakedef.h"
+
+#include <nds.h>
+
+typedef struct
+{
+	u32	button;
+	int	game_key;	// bound by default.cfg
+	int	menu_key;
+} buttonmap_t;
+
+static const buttonmap_t buttonmap[] =
+{
+	{KEY_UP,	K_UPARROW,	K_UPARROW},	// +forward
+	{KEY_DOWN,	K_DOWNARROW,	K_DOWNARROW},	// +back
+	{KEY_LEFT,	K_LEFTARROW,	K_LEFTARROW},	// +left
+	{KEY_RIGHT,	K_RIGHTARROW,	K_RIGHTARROW},	// +right
+	{KEY_A,		K_CTRL,		K_ENTER},	// +attack
+	{KEY_B,		K_SPACE,	K_ESCAPE},	// +jump
+	{KEY_X,		'/',		K_BACKSPACE},	// impulse 10 (next weapon)
+	{KEY_Y,		K_SHIFT,	K_SHIFT},	// +speed
+	{KEY_L,		',',		K_PGUP},	// +moveleft
+	{KEY_R,		'.',		K_PGDN},	// +moveright
+	{KEY_START,	K_ESCAPE,	K_ESCAPE},	// togglemenu
+};
+
+#define NUM_BUTTONS	(sizeof(buttonmap) / sizeof(buttonmap[0]))
+
+// The key sent for each held button, released even when key_dest changed
+static int	held_key[NUM_BUTTONS];
+
+static qboolean	touching;
+static int	touch_key;		// held by the stylus
+
+void IN_Init (void)
+{
+}
+
+void IN_Shutdown (void)
+{
+}
+
+void Sys_SendKeyEvents (void)
+{
+	u32		down, up;
+	int		i;
+	touchPosition	touch;
+
+	// sleeps while the lid is closed, quits on the DSi power button or
+	// L+R+START+SELECT, keeping the config like the quit command
+	if (!pmMainLoop ())
+	{
+		Host_Shutdown ();
+		Sys_Quit ();
+	}
+	scanKeys ();
+	down = keysDown ();
+	up = keysUp ();
+
+	for (i = 0; i < NUM_BUTTONS; i++)
+	{
+		if ((up & buttonmap[i].button) && held_key[i])
+		{
+			Key_Event (held_key[i], false);
+			held_key[i] = 0;
+		}
+		if (down & buttonmap[i].button)
+		{
+			held_key[i] = key_dest == key_game ? buttonmap[i].game_key : buttonmap[i].menu_key;
+			Key_Event (held_key[i], true);
+		}
+	}
+
+	if (keysHeld () & KEY_TOUCH)
+	{
+		touchRead (&touch);
+		if (!touching)
+		{
+			// a new touch selects a weapon, in 2D coordinates
+			touch_key = Sbar_TouchKey (touch.px * vid.width / 256, touch.py * vid.height / 192);
+			if (touch_key)
+				Key_Event (touch_key, true);
+		}
+		touching = true;
+	}
+	else if (touching)
+	{
+		if (touch_key)
+			Key_Event (touch_key, false);
+		touch_key = 0;
+		touching = false;
+	}
+}
+
+void IN_Commands (void)
+{
+}
+
+void IN_Move (usercmd_t *cmd)
+{
+}
